### ISSUE TITLE
twistd does not give up Group 0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10366]
|<img alt="thomaswaldmann's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| thomaswaldmann reported|
|-|-|
|Trac ID|trac#521|
|Type|enhancement|
|Created|2004-02-24 00:55:16Z|

<details><summary>Searchable metadata</summary>

```
trac-id__521 521
type__enhancement enhancement
reporter__thomaswaldmann thomaswaldmann
priority__lowest lowest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1077584116000000 1077584116000000
changetime__1472150025715599 1472150025715599
version__ 
owner__ 
cc__exarkun cc__itamarst cc__moshez cc__jknight cc__thomaswaldmann
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="thomaswaldmann's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomaswaldmann commented|
|-|-|

```
#!html
<pre>
Important part of the code:
...
uid = pwd.getpwnam('whatever')[2]
gid = grp.getgrnam('whatever')[2]
application = app.Application("web", uid=uid, gid=gid)
application.listenTCP(80, site, interface='1.2.3.4')

# cat /proc/&lt;pid>/status:
Name:   twistd2.3
State:  S (sleeping)
Tgid:   27794
Pid:    27794
PPid:   30459
TracerPid:      0
Uid:    20003   20003   20003   20003
Gid:    20003   20003   20003   20003
FDSize: 256
Groups: 0    &lt;------ I think this should not be there.
...

$ twistd2.3 --version
twistd (the Twisted daemon) 1.1.1
Copyright (c) 2000-2003 Matthew William Lefkowitz, all rights reserved.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
twisted.internet.app is deprecated. New package is twisted.application (you
probably want to check code in twisted.application.app to see if it does or does
not have this issue.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
exarkun mentioned that twisted.internet.process.set_uid has code that does the
right thing. Or something.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
twistd now uses the same setuid code as spawnProcess(), which was and still is
correct.

How to unit test this effectively isn't immediately clear to me.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
setgroups() fails unconditionally when called by a process without the
appropriate posix capability (or, on systems without posix capabilities, UID 0 I
guess), even if it would set the groups to what they are already.

I added a try/except around the call which silently discards EPERM OSErrors. 
I'm not sure if this is the best behavior.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
"silently discards EPERM OSErrors"

That sounds evil. It should not ignore errors like that lest you somehow end up accidently with more 
permissions than you expected/wanted. 

However, if it verifies after the call that the current group membership (os.getgroups()) does not contain 
more groups than were asked for, then an error from setgroups() can be safely ignored.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Changed title. t.i.app.Application is irrelevant and deprecated.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
EPERM is now only discarded if the groups the process belongs to include no
groups which were not specified by the setgroups() call.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc commented|
|-|-|

In changeset b43ebc3999b3351dfb8e33e97371ca40654344c8
```
#!CommitTicketReference repository="" revision="b43ebc3999b3351dfb8e33e97371ca40654344c8"
Merge pull request #521 from twisted/8780-iocp-ipv6

Author: rodrigc
Reviewer: opalmer
Fixes: #8780

Fix AcceptEx() test on Windows
```
